File: Lexer/Typescript/lexer.ts
import { Token, TokenType, newToken, IDENTIFIER_MAP } from "./token";

export class Lexer {
  private input: string;
  private ch: string;
  private curr: number;
  private peek: number;

  constructor (input: string) {
    this.input = input;
    this.ch = '\0';
    this.curr = 0;
    this.peek = 0;
    this.readChar();
  }

  public nextToken(): Token {
    let tok: Token = newToken(TokenType.EOF, "\0");

    while (isWhitespace(this.ch))
      this.readChar();

    switch (this.ch) {
      case '\0': return tok;
      case '*': tok = newToken(TokenType.ASTERISK, "*"); break;
      case '-': tok = newToken(TokenType.DASH, "-"); break;
      case '+': tok = newToken(TokenType.PLUS, "+"); break;
      case '=': tok = newToken(TokenType.EQUAL, "="); break;
      case '<': tok = newToken(TokenType.LESSTHAN, "<"); break;
      case '>': tok = newToken(TokenType.GREATERTHAN, ">"); break;
      case '/': tok = newToken(TokenType.SLASH, "/"); break;
      case ';': tok = newToken(TokenType.SEMICOLON, ";"); break;
      case ':': tok = newToken(TokenType.COLON, ";"); break;
      case '(': tok = newToken(TokenType.LPAREN, "("); break;
      case ')': tok = newToken(TokenType.RPAREN, ")"); break;
      case '{': tok = newToken(TokenType.LBRACE, "{"); break;
      case '}': tok = newToken(TokenType.RBRACE, "}"); break;
      default:
        if (isLetter(this.ch)) {
          const str = this.readIdentifier();
          const ttype = IDENTIFIER_MAP[str];
          return ttype == null || ttype == undefined ?
            newToken(TokenType.IDENTIFIER, str) :
            newToken(ttype, str);
        } else if (isDigit(this.ch))
          return newToken(TokenType.NUMBER, this.readNumber());
        else return newToken(TokenType.ILLEGAL, "ILLEGAL");
    }
    this.readChar();
    return tok;
  }

  private readIdentifier(): string {
    const pos = this.curr;
    while (isLetter(this.ch) || this.ch == '_')
      this.readChar();
    return this.input.substring(pos, this.curr);
  }

  private readNumber(): string {
    const pos = this.curr;
    while (isDigit(this.ch))
      this.readChar();
    return this.input.substring(pos, this.curr);
  }

  private readChar() {
    if (this.peek >= this.input.length)
      this.ch = '\0';
    else
      this.ch = this.input[this.peek];
    this.curr = this.peek++;
  }
}

function isLetter(str: string): boolean {
  return /^[a-xA-Z]+$/.test(str);
}

function isDigit(str: string): boolean {
  return /^[0-9]+$/.test(str);
}

function isWhitespace(str: string): boolean {
  return /^[\r\n\t\s]+$/.test(str);
}

File: Lexer/Lua/lexer.lua
---@class Lexer
---@field input string
---@field char string
---@field curr number
---@field peek number
local Lexer = {
    input = "",
    char = "",
    curr = 1,
    peek = 1,
}

local token = require("token")
local newToken = token.newToken
local TokenType = token.TokenType
local IdentifierMap = token.IdentifierMap

---@return Token
function Lexer:nextToken()
    while self.char:match("%s") do
        self:readChar()
    end

    local tok
    local c = self.char
    if c == "\0" then
        tok = newToken(TokenType.EOF, "\0")
    elseif c == "*" then
        tok = newToken(TokenType.ASTERISK, "*")
    elseif c == "/" then
        tok = newToken(TokenType.SLASH, "/")
    elseif c == "-" then
        tok = newToken(TokenType.DASH, "-")
    elseif c == "+" then
        tok = newToken(TokenType.PLUS, "+")
    elseif c == "=" then
        tok = newToken(TokenType.EQUAL, "=")
    elseif c == "<" then
        tok = newToken(TokenType.LESSTHAN, "<")
    elseif c == ">" then
        tok = newToken(TokenType.GREATERTHAN, ">")
    elseif c == ";" then
        tok = newToken(TokenType.SEMICOLON, ";")
    elseif c == ":" then
        tok = newToken(TokenType.COLON, ":")
    elseif c == "(" then
        tok = newToken(TokenType.LPAREN, "(")
    elseif c == ")" then
        tok = newToken(TokenType.RPAREN, ")")
    elseif c == "{" then
        tok = newToken(TokenType.LBRACE, "{")
    elseif c == "}" then
        tok = newToken(TokenType.RBRACE, "}")
    else
        if self.char:match("[a-zA-Z+]") then
            local ident = self:readIdentifier()
            local ttype = IdentifierMap[ident]
            if ttype ~= nil then
                return newToken(ttype, ident)
            else
                return newToken(TokenType.IDENTIFIER, ident)
            end
        elseif self.char:match("[0-9]") then
            return newToken(TokenType.NUMBER, self:readNumber())
        end
        return newToken(TokenType.ILLEGAL, "ILLEGAL")
    end
    self:readChar()

    return tok
end

---@return string
function Lexer:readIdentifier()
    local pos = self.curr
    while self.char:match("[a-zA-Z_]") do
        self:readChar()
    end
    return self.input:sub(pos, self.curr - 1)
end

---@return string
function Lexer:readNumber()
    local pos = self.curr
    while self.char:match("[0-9]") do
        self:readChar()
    end
    return self.input:sub(pos, self.curr - 1)
end

---@return nil
function Lexer:readChar()
    if self.peek > self.input:len() then
        self.char = "\0"
    else
        self.char = self.input:sub(self.peek, self.peek)
    end
    self.curr = self.peek
    self.peek = self.peek + 1
end

return Lexer
